## 단원별 심화 연습 문제

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import seaborn as sns
import glob

# set floating point formatting
pd.options.display.float_format = '{:,.1f}'.format

### Q4

**범위**
- (이전 범위 포함)
- `groupby()`, `pivot_table()`

In [2]:
# 데이터셋
taxis = sns.load_dataset('taxis')
taxis.head()

pickup             dropoff  passengers  distance  fare  tip  \
0 2019-03-23 20:21:09 2019-03-23 20:27:24           1       1.6   7.0  2.1   
1 2019-03-04 16:11:55 2019-03-04 16:19:00           1       0.8   5.0  0.0   
2 2019-03-27 17:53:01 2019-03-27 18:00:25           1       1.4   7.5  2.4   
3 2019-03-10 01:23:59 2019-03-10 01:49:51           1       7.7  27.0  6.2   
4 2019-03-30 13:27:42 2019-03-30 13:37:14           3       2.2   9.0  1.1   

   tolls  total   color      payment            pickup_zone  \
0    0.0   12.9  yellow  credit card        Lenox Hill West   
1    0.0    9.3  yellow         cash  Upper West Side South   
2    0.0   14.2  yellow  credit card          Alphabet City   
3    0.0   37.0  yellow  credit card              Hudson Sq   
4    0.0   13.4  yellow  credit card           Midtown East   

            dropoff_zone pickup_borough dropoff_borough  
0    UN/Turtle Bay South      Manhattan       Manhattan  
1  Upper West Side South      Manhattan       Manhattan  
2           West Village      Manhattan       Manhattan  
3         Yorkville West      Manhattan       Manhattan  
4         Yorkville West      Manhattan       Manhattan

`taxis` 데이터에서 요일별 `distance`의 평균을 산출하세요
(DataFrame 형식으로 출력하세요)

- [참고] 0: 월요일 ~ 6: 일요일

In [11]:
# 코드를 입력해 주세요
taxis['dayofweek'] = taxis['pickup'].dt.dayofweek

taxis.groupby('dayofweek')['distance'].mean().reset_index()

dayofweek  distance
0          0       3.2
1          1       3.0
2          2       3.1
3          3       3.0
4          4       2.9
5          5       2.9
6          6       3.0

<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>dayofweek</th>
      <th>distance</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>3.2</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2</td>
      <td>3.1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>3</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>4</td>
      <td>2.9</td>
    </tr>
    <tr>
      <th>5</th>
      <td>5</td>
      <td>2.9</td>
    </tr>
    <tr>
      <th>6</th>
      <td>6</td>
      <td>3.0</td>
    </tr>
  </tbody>
</table>
</div>

위에서 산출한 통계를 바탕으로 다음을 도출합니다.

- 요일별 `distance`의 평균 값을 구하였습니다.
- 구한 평균치보다 작은(미만) 인 데이터만 추출하세요.
- (예시) 0: 월요일 평균 `distance`가 `3.215971`이라면 `3.215971`보다 작은 데이터만 추출합니다. (월~ 일요일까지 모두 적용)
- 필터 후 `distance`, `fare`, `tip`에 대한 중앙값(median)과 표준편차(std)를 산출합니다.

In [12]:
# 코드를 입력해 주세요
sample = taxis.groupby('dayofweek')['distance'].mean().reset_index()

def is_small(x):
    return x['distance'] < sample.loc[x['dayofweek'], 'distance']

cond1 = taxis.apply(is_small, axis=1)

taxis.loc[cond1, ['distance', 'fare', 'tip']].agg(['median', 'std'])

distance  fare  tip
median       1.3   7.5  1.6
std          0.7   4.8  1.5

<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>distance</th>
      <th>fare</th>
      <th>tip</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>median</th>
      <td>1.2</td>
      <td>7.5</td>
      <td>1.6</td>
    </tr>
    <tr>
      <th>std</th>
      <td>0.7</td>
      <td>4.7</td>
      <td>1.4</td>
    </tr>
  </tbody>
</table>
</div>

다음의 피벗테이블을 완성해 주세요
- 요일(`dayofweek`)과 `payment` 별 `fare`의 중앙값(median)과 표준편차(std)를 산출합니다.

In [13]:
# 코드를 입력해 주세요
taxis.pivot_table(index='dayofweek', columns='payment', values='fare', aggfunc=['median', 'std'])

median              std            
payment     cash credit card cash credit card
dayofweek                                    
0            8.5         9.5 12.8        12.4
1            8.0        10.0 11.9        12.9
2            8.5         9.5 13.2        11.2
3            8.5        10.0  9.5        11.1
4            8.5         9.5 10.3        12.1
5            9.0         9.2  8.6        10.9
6            8.0         9.5 12.9        10.6

<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead tr th {
        text-align: left;
    }

    .dataframe thead tr:last-of-type th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr>
      <th></th>
      <th colspan="2" halign="left">median</th>
      <th colspan="2" halign="left">std</th>
    </tr>
    <tr>
      <th>payment</th>
      <th>cash</th>
      <th>credit card</th>
      <th>cash</th>
      <th>credit card</th>
    </tr>
    <tr>
      <th>dayofweek</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>8.8</td>
      <td>9.5</td>
      <td>13.3</td>
      <td>12.4</td>
    </tr>
    <tr>
      <th>1</th>
      <td>8.0</td>
      <td>10.0</td>
      <td>12.4</td>
      <td>12.4</td>
    </tr>
    <tr>
      <th>2</th>
      <td>8.5</td>
      <td>9.5</td>
      <td>10.1</td>
      <td>11.2</td>
    </tr>
    <tr>
      <th>3</th>
      <td>8.5</td>
      <td>10.0</td>
      <td>9.2</td>
      <td>11.3</td>
    </tr>
    <tr>
      <th>4</th>
      <td>8.0</td>
      <td>9.5</td>
      <td>8.6</td>
      <td>11.9</td>
    </tr>
    <tr>
      <th>5</th>
      <td>9.0</td>
      <td>9.5</td>
      <td>8.6</td>
      <td>11.0</td>
    </tr>
    <tr>
      <th>6</th>
      <td>8.0</td>
      <td>9.5</td>
      <td>13.3</td>
      <td>11.0</td>
    </tr>
  </tbody>
</table>
</div>

## 제출

제출을 위해 새로 로드된 택시 데이터셋에서 요일별 `distance`의 평균을 계산한 결과를 DataFrame 형식으로 `result_df`에 저장하세요.

In [20]:
taxis = sns.load_dataset('taxis')
taxis['dayofweek'] = taxis['pickup'].dt.dayofweek
result_df = taxis.groupby('dayofweek')['distance'].mean().reset_index()

result_df

dayofweek  distance
0          0       3.2
1          1       3.0
2          2       3.1
3          3       3.0
4          4       2.9
5          5       2.9
6          6       3.0

## 채점 수행

아래 코드는 채점 수행을 위한 코드입니다.

따라서 이를 수정했을시 **채점이 제대로 이루어지지 않습니다.**

**주의**: 채점 코드를 실행하기 전에 반드시 코드 파일을 한번 저장하시길 바랍니다.

In [21]:
import os

result_df.to_json('result.json')

os.system('elice_grade result.json cds_ds_exercise.ipynb')

send files ['result.json', 'cds_ds_exercise.ipynb'] for grade...
waiting result...
waiting result...
waiting result...
done!

Score: 100.000000
Duration: 2.588 seconds
=== Message ===
제출 완료되었습니다.


0